server/athenian/api/typing_utils.py

```diff
@@ -365,47 +373,77 @@ def df_from_structs(
         except StopIteration:
             return pd.DataFrame()
         assert isinstance(first_item, NumpyStruct)
-        dtype = first_item.dtype
-        nested_fields = first_item.nested_dtypes
-        coerced_datas = [first_item.coerced_data]
-        for k, v in first_item.items():
-            if k not in dtype.names or k in nested_fields:
-                columns[k] = [v]
+        inspector = _NumpyStructInspector(first_item, None)
+        dc_items = inspector.dataclasses_columns.items()
+        indirect_items = inspector.indirect_columns.items()
+        coerced_datas = inspector.coerced_datas
+        columns = inspector.columns
         for item in items_iter:
             coerced_datas.append(item.coerced_data)
-            for k in columns:
+
+            for k in inspector.direct_columns:
                 columns[k].append(getattr(item, k))
-        table_array = np.frombuffer(b"".join(coerced_datas), dtype=dtype)
+
+            for k, (field, subfield) in dc_items:
+                columns[k].append(getattr(item[field], subfield))
+
+            for k, sibling_keys in indirect_items:
+                sibling_ns = getattr(item, k)
+                coerced_datas.append(sibling_ns.coerced_data)
+                for sibling_k, full_k in sibling_keys:
+                    columns[full_k].append(getattr(sibling_ns, sibling_k))
+
+        table_array = np.frombuffer(b"".join(coerced_datas), dtype=inspector.dtype)
         del coerced_datas
     else:
         items_iter = iter(items)
         try:
             first_item = next(items_iter)
         except StopIteration:
             return pd.DataFrame()
+
         assert isinstance(first_item, NumpyStruct)
-        dtype = first_item.dtype
-        nested_fields = first_item.nested_dtypes
-        itemsize = dtype.itemsize
-        coerced_datas = bytearray(itemsize * length)
-        coerced_datas[:itemsize] = first_item.coerced_data
-        for k, v in first_item.items():
-            if k not in dtype.names or k in nested_fields:
-                columns[k] = column = [None] * length
-                column[0] = v
+        inspector = _NumpyStructInspector(first_item, length)
+
+        coerced_pos = inspector.dtype.itemsize
+        coerced_datas_buf = inspector.get_coerced_datas_buf(length)
+        columns = inspector.columns
+
+        dc_items = inspector.dataclasses_columns.items()
+        indirect_items = inspector.indirect_columns.items()
         for i, item in enumerate(items_iter, 1):
-            coerced_datas[i * itemsize : (i + 1) * itemsize] = item.coerced_data
-            for k in columns:
+            coerced_data = item.coerced_data
+            coerced_datas_buf[coerced_pos : coerced_pos + len(coerced_data)] = coerced_data
```

Comment:
```suggestion
            coerced_datas_buf[coerced_pos : coerced_pos + itemsize] = coerced_data
```
(save a few instructions)